In [3]:
import pandas as pd
import os

raw_path = "../data/raw"

os.listdir(raw_path)

['taxi_trip_data.csv',
 'taxi_zone_geo.csv',
 '.complete',
 'original_cleaned_nyc_taxi_data_2018.csv']

In [4]:
#Loading the taxi trips .csv
trips = pd.read_csv("../data/raw/taxi_trip_data.csv", nrows=100_000)

#Loading the geographics zones .csv
zones = pd.read_csv("../data/raw/taxi_zone_geo.csv")

#Loading "clean"  data
cleaned = pd.read_csv("../data/raw/original_cleaned_nyc_taxi_data_2018.csv", nrows=100_000)

# Quick Inspecion

In [5]:
def quick(df):
    print("shape:", df.shape)
    print("\ncolumns:")
    print(df.columns)
    print("\ndtypes:")
    print(df.dtypes)
    print("\nisnull:")
    print(df.isnull().sum())
    display(df.head())

## Trips

In [6]:
quick(trips)

shape: (100000, 17)

columns:
Index(['vendor_id', 'pickup_datetime', 'dropoff_datetime', 'passenger_count',
       'trip_distance', 'rate_code', 'store_and_fwd_flag', 'payment_type',
       'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'imp_surcharge', 'total_amount', 'pickup_location_id',
       'dropoff_location_id'],
      dtype='object')

dtypes:
vendor_id                int64
pickup_datetime         object
dropoff_datetime        object
passenger_count          int64
trip_distance          float64
rate_code                int64
store_and_fwd_flag      object
payment_type             int64
fare_amount            float64
extra                  float64
mta_tax                float64
tip_amount             float64
tolls_amount           float64
imp_surcharge          float64
total_amount           float64
pickup_location_id       int64
dropoff_location_id      int64
dtype: object

isnull:
vendor_id              0
pickup_datetime        0
dropoff_datetime     

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
0,2,2018-03-29 13:37:13,2018-03-29 14:17:01,1,18.15,3,N,1,70.0,0.0,0.0,16.16,10.50,0.3,96.96,161,1
1,2,2018-03-29 13:37:18,2018-03-29 14:15:33,1,4.59,1,N,1,25.0,0.0,0.5,5.16,0.00,0.3,30.96,13,230
2,2,2018-03-29 13:26:57,2018-03-29 13:28:03,1,0.30,1,N,1,3.0,0.0,0.5,0.76,0.00,0.3,4.56,231,231
3,2,2018-03-29 13:07:48,2018-03-29 14:03:05,2,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,0.3,61.67,231,138
4,2,2018-03-29 14:19:11,2018-03-29 15:19:59,5,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,0.3,62.47,87,138


## Zones

In [7]:
quick(zones)

shape: (263, 4)

columns:
Index(['zone_id', 'zone_name', 'borough', 'zone_geom'], dtype='object')

dtypes:
zone_id       int64
zone_name    object
borough      object
zone_geom    object
dtype: object

isnull:
zone_id      0
zone_name    0
borough      0
zone_geom    0
dtype: int64


,zone_id,zone_name,borough,zone_geom
0,1,Newark Airport,EWR,"POLYGON((-74.1856319999999 40.6916479999999, -..."
1,3,Allerton/Pelham Gardens,Bronx,"POLYGON((-73.848596761 40.8716707849999, -73.8..."
2,18,Bedford Park,Bronx,"POLYGON((-73.8844286139999 40.8668003789999, -..."
3,20,Belmont,Bronx,"POLYGON((-73.8839239579998 40.8644177609999, -..."
4,31,Bronx Park,Bronx,"POLYGON((-73.8710017319999 40.8572767429999, -..."


## Cleaned

In [8]:
quick(cleaned)

shape: (100000, 21)

columns:
Index(['Unnamed: 0', 'trip_distance', 'rate_code', 'store_and_fwd_flag',
       'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount',
       'tolls_amount', 'imp_surcharge', 'total_amount', 'pickup_location_id',
       'dropoff_location_id', 'year', 'month', 'day', 'day_of_week',
       'hour_of_day', 'trip_duration', 'calculated_total_amount'],
      dtype='object')

dtypes:
Unnamed: 0                   int64
trip_distance              float64
rate_code                    int64
store_and_fwd_flag          object
payment_type                 int64
fare_amount                float64
extra                      float64
mta_tax                    float64
tip_amount                 float64
tolls_amount               float64
imp_surcharge              float64
total_amount               float64
pickup_location_id           int64
dropoff_location_id          int64
year                         int64
month                        int64
day                 

,Unnamed: 0,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,...,total_amount,pickup_location_id,dropoff_location_id,year,month,day,day_of_week,hour_of_day,trip_duration,calculated_total_amount
0,3,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,...,61.67,231,138,2018,3,29,3,13,3317.0,61.67
1,4,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,...,62.47,87,138,2018,3,29,3,14,3648.0,62.47
2,5,11.60,1,N,1,42.0,0.0,0.5,14.57,5.76,...,63.13,68,138,2018,3,29,3,14,3540.0,63.13
3,10,5.10,1,N,1,26.5,1.0,0.5,5.65,0.00,...,33.95,186,33,2018,3,29,3,16,2585.0,33.95
4,12,11.11,1,N,1,45.5,1.0,0.5,10.61,5.76,...,63.67,163,138,2018,3,29,3,16,4521.0,63.67


## General view

In [9]:
datasets = {
    "trips": trips,
    "zones": zones,
    "cleaned": cleaned
}

for name, df in datasets.items():
    print(f"\n===== {name} =====")
    print("Shape:", df.shape)
    print("Columns:", df.columns.tolist())
    print("\nTypes:")
    print(df.dtypes)


===== trips =====
Shape: (100000, 17)
Columns: ['vendor_id', 'pickup_datetime', 'dropoff_datetime', 'passenger_count', 'trip_distance', 'rate_code', 'store_and_fwd_flag', 'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'imp_surcharge', 'total_amount', 'pickup_location_id', 'dropoff_location_id']

Types:
vendor_id                int64
pickup_datetime         object
dropoff_datetime        object
passenger_count          int64
trip_distance          float64
rate_code                int64
store_and_fwd_flag      object
payment_type             int64
fare_amount            float64
extra                  float64
mta_tax                float64
tip_amount             float64
tolls_amount           float64
imp_surcharge          float64
total_amount           float64
pickup_location_id       int64
dropoff_location_id      int64
dtype: object

===== zones =====
Shape: (263, 4)
Columns: ['zone_id', 'zone_name', 'borough', 'zone_geom']

Types:
zone_id       int64

We are choosing the "taxi_trip_data" as our main source of information because is the .csv that only has information about the taxi trips, and that is what we are going to analyse in this project

# EDA - taxi_trip_data.csv

## Seeing the data 'estructure'

In [1]:
import pandas as pd

#Loading the taxi trips .csv
trips = pd.read_csv("../data/raw/taxi_trip_data.csv", nrows=100_000)

In [2]:
trips.shape

(100000, 17)

In [3]:
trips.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
0,2,2018-03-29 13:37:13,2018-03-29 14:17:01,1,18.15,3,N,1,70.0,0.0,0.0,16.16,10.50,0.3,96.96,161,1
1,2,2018-03-29 13:37:18,2018-03-29 14:15:33,1,4.59,1,N,1,25.0,0.0,0.5,5.16,0.00,0.3,30.96,13,230
2,2,2018-03-29 13:26:57,2018-03-29 13:28:03,1,0.30,1,N,1,3.0,0.0,0.5,0.76,0.00,0.3,4.56,231,231
3,2,2018-03-29 13:07:48,2018-03-29 14:03:05,2,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,0.3,61.67,231,138
4,2,2018-03-29 14:19:11,2018-03-29 15:19:59,5,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,0.3,62.47,87,138


In [4]:
trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   vendor_id            100000 non-null  int64  
 1   pickup_datetime      100000 non-null  object 
 2   dropoff_datetime     100000 non-null  object 
 3   passenger_count      100000 non-null  int64  
 4   trip_distance        100000 non-null  float64
 5   rate_code            100000 non-null  int64  
 6   store_and_fwd_flag   100000 non-null  object 
 7   payment_type         100000 non-null  int64  
 8   fare_amount          100000 non-null  float64
 9   extra                100000 non-null  float64
 10  mta_tax              100000 non-null  float64
 11  tip_amount           100000 non-null  float64
 12  tolls_amount         100000 non-null  float64
 13  imp_surcharge        100000 non-null  float64
 14  total_amount         100000 non-null  float64
 15  pickup_location_id

"pickup_datetime" and "dropoff_datetime" as object is "wrong" i must change the type for 'date' in the pipeline

payment_type as a number, i found in the description of the dataset the equivalent numbers, so in the future i will need to create a auxiliary table

rate_code is categoric besides beeing int

store_and_fwd_flg is a object but is categoric

In [8]:
schema = pd.DataFrame({
    "column": trips.columns,
    "dtype": trips.dtypes.values,
    "null_count": trips.isnull().sum().values,
    "unique_count": trips.nunique().values
})

schema

,column,dtype,null_count,unique_count
0,vendor_id,int64,0,3
1,pickup_datetime,object,0,99618
2,dropoff_datetime,object,0,99572
3,passenger_count,int64,0,10
4,trip_distance,float64,0,2836
5,rate_code,int64,0,5
6,store_and_fwd_flag,object,0,2
7,payment_type,int64,0,4
8,fare_amount,float64,0,625
9,extra,float64,0,13


17 colunas
100.000 registros na amostra
0 nulos em todas as colunas da amostra
8 float64
6 int64
3 object

## Seeing the data quality

In [9]:
trips.isnull().sum()

vendor_id              0
pickup_datetime        0
dropoff_datetime       0
passenger_count        0
trip_distance          0
rate_code              0
store_and_fwd_flag     0
payment_type           0
fare_amount            0
extra                  0
mta_tax                0
tip_amount             0
tolls_amount           0
imp_surcharge          0
total_amount           0
pickup_location_id     0
dropoff_location_id    0
dtype: int64

In [10]:
trips.duplicated().sum()

np.int64(134)

In [11]:
trips['payment_type'].value_counts()

payment_type
1    95188
2     4434
3      291
4       87
Name: count, dtype: int64

In [12]:
trips["vendor_id"].value_counts()

vendor_id
2    60283
1    39250
4      467
Name: count, dtype: int64

In [13]:
trips["rate_code"].value_counts()

rate_code
1    95272
5     2358
3     1889
4      473
2        8
Name: count, dtype: int64

We found 134 duplicated lines, but this not mean that this 134 is incorrect data, we need to see the content of this lines

In [17]:
duplicates = trips[trips.duplicated(keep=False)]
#keep = false, show from the first apparence to the final, if we keep=True, will only show from the second ocorrence

In [15]:
duplicates.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
1314,1,2018-03-13 16:52:05,2018-03-13 17:32:14,1,8.20,1,N,1,30.0,1.0,0.5,6.35,0.00,0.3,38.15,50,17
2724,1,2018-03-31 20:40:55,2018-03-31 21:02:33,1,6.70,1,N,1,22.0,0.5,0.5,4.65,0.00,0.3,27.95,264,82
3118,2,2018-03-13 23:17:34,2018-03-13 23:34:06,1,9.38,1,N,1,27.0,0.5,0.5,4.00,0.00,0.3,32.30,138,144
3216,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264
3249,2,2018-03-22 23:04:46,2018-03-22 23:35:57,5,16.86,1,N,1,47.5,0.5,0.5,7.32,0.00,0.3,56.12,132,255


In [16]:
duplicates.shape

(268, 17)

In [20]:
duplicates.sort_values(
    by=["pickup_datetime", "dropoff_datetime"]
).head(10)

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
6777,2,2018-03-01 00:53:19,2018-03-01 01:09:54,2,7.33,1,N,1,22.0,0.5,0.5,4.66,0.00,0.3,27.96,209,263
28936,2,2018-03-01 00:53:19,2018-03-01 01:09:54,2,7.33,1,N,1,22.0,0.5,0.5,4.66,0.00,0.3,27.96,209,263
4754,2,2018-03-01 00:58:35,2018-03-01 01:25:34,1,19.83,1,N,1,52.5,0.5,0.5,9.00,0.00,0.3,62.80,132,49
27911,2,2018-03-01 00:58:35,2018-03-01 01:25:34,1,19.83,1,N,1,52.5,0.5,0.5,9.00,0.00,0.3,62.80,132,49
8858,2,2018-03-01 03:47:33,2018-03-01 03:48:28,2,0.31,1,N,1,3.0,0.5,0.5,1.08,0.00,0.3,5.38,137,137
33095,2,2018-03-01 03:47:33,2018-03-01 03:48:28,2,0.31,1,N,1,3.0,0.5,0.5,1.08,0.00,0.3,5.38,137,137
25690,2,2018-03-01 16:24:07,2018-03-01 17:19:05,1,16.57,1,N,1,50.5,1.0,0.5,11.61,5.76,0.3,69.67,138,13
49291,2,2018-03-01 16:24:07,2018-03-01 17:19:05,1,16.57,1,N,1,50.5,1.0,0.5,11.61,5.76,0.3,69.67,138,13
3216,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264
41226,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264


I identified 134 duplicated lines, we can see that all columns are duplicates so probably are really duplicates but to have more confidence i will run another lines to analyse especifically other columns

In [22]:
duplicates["vendor_id"].value_counts()

vendor_id
2    166
1    102
Name: count, dtype: int64

In [23]:
duplicates[[
    "vendor_id",
    "pickup_datetime",
    "dropoff_datetime",
    "trip_distance",
    "fare_amount",
    "pickup_location_id",
    "dropoff_location_id"
]]

,vendor_id,pickup_datetime,dropoff_datetime,trip_distance,fare_amount,pickup_location_id,dropoff_location_id
1314,1,2018-03-13 16:52:05,2018-03-13 17:32:14,8.20,30.0,50,17
2724,1,2018-03-31 20:40:55,2018-03-31 21:02:33,6.70,22.0,264,82
3118,2,2018-03-13 23:17:34,2018-03-13 23:34:06,9.38,27.0,138,144
3216,1,2018-03-01 17:53:05,2018-03-01 19:15:23,15.30,56.5,264,264
3249,2,2018-03-22 23:04:46,2018-03-22 23:35:57,16.86,47.5,132,255
...,...,...,...,...,...,...,...
99218,2,2018-03-06 10:10:41,2018-03-06 11:08:08,13.35,48.0,138,48
99425,2,2018-03-06 16:42:33,2018-03-06 16:43:00,0.00,65.0,234,234
99517,2,2018-03-07 07:15:15,2018-03-07 07:39:43,8.56,27.5,138,236
99630,2,2018-03-08 13:05:38,2018-03-08 13:40:09,15.53,44.5,132,145


Running another checkup i can confirm that is really duplicates, because we can see that ALL columns are duplicates too

In the 100.000 we found 134 duplicates that translate to 268 lines

So in the pipeline this duplicates will be removed

In [26]:
duplicate_rows = len(duplicates)
percentage = duplicate_rows / len(trips) * 100

print(f"Rows involved in duplication: {duplicate_rows}")
print(f"Percentage: {percentage:.3f}%")

Rows involved in duplication: 268
Percentage: 0.268%


In [25]:
duplicate_counts = trips.value_counts()

duplicate_counts[duplicate_counts > 1].head(20)

vendor_id  pickup_datetime      dropoff_datetime     passenger_count  trip_distance  rate_code  store_and_fwd_flag  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  imp_surcharge  total_amount  pickup_location_id  dropoff_location_id
2          2018-03-07 07:15:15  2018-03-07 07:39:43  2                8.56           1          N                   1             27.5         0.0    0.5      6.81        5.76          0.3            40.87         138                 236                    2
           2018-03-02 20:56:02  2018-03-02 21:47:55  1                25.40          1          N                   1             69.5         0.5    0.5      18.24       2.16          0.3            91.20         90                  201                    2
1          2018-03-17 15:20:02  2018-03-17 16:06:17  2                5.30           1          N                   1             29.5         0.0    0.5      9.05        0.00          0.3            39.35         79            

## Time series analysis